File: mc2csv.php
<?php
// This software is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// You should have received a copy of the GNU General Public License
// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.

/**
 * Takes a Moodle XML export file of questions and exports all the
 * multiple choice questions into a csv file.
 * 
 * Usage php mc2csv.php --in questions.xml
 *
 * Optional arguments:
 * --delimiter char    Delimiter character for the csv file, default is ";"
 * --keephtml          Do not strip html tags at all.
 * --keeptags 'a,b,c'  List of tags that should not be eliminated when
 *                     stripping html from the questions. The default
 *                     tags that are not purged: i,img,s,strong,sub,sub,u
 * --out filename.csv  if not provided the base name of the xml file is
 *                     used, trailed by a .csv suffix.
 *         
 **/

// Predefined variables.
$infile = $outfile = null;
$delimiter = ';';
$keeptags = ['i', 'img', 's', 'strong', 'sub', 'sub', 'u'];
$striptags = true;

$args = $_SERVER['argv'];
array_shift($args);

// Start handling the command line arguments.
$carg = null;
while ($arg = array_shift($args)) {
    if ($arg === '--in' || $arg === '--out' || $arg === '--keeptags' || $arg === '--delimiter') {
        $carg = $arg;
        continue;
    }
    if ($arg === '--keephtml') {
        $striptags = false;
        continue;
    }
    if ($arg === '--help') {
        $echo = false;
        $fp = fopen($_SERVER['argv'][0], 'r');
        while ($line = fgets($fp)) {
            if ($echo) {
                if (trim($line) === '**/') {
                    fclose($fp);
                    dieNice();
                }
                $line = substr($line, 3);
                if (!empty($line)) {
                    echo $line;
                } else {
                    echo "\n";
                }
                continue;
            }
            if (substr($line, 0, 3) === '/**') {
                $echo = true;
            }
        }
        dieNice();
    }
    if ($carg === '--in') {
        $infile = $arg;
        $carg = null;
        continue;
    }
    if ($carg === '--out') {
        $outfile = $arg;
        $carg = null;
        continue;
    }
    if ($carg === '--keeptags') {
        $keeptags = array_map(fn($t) => trim($t), explode(',', $arg));
        $carg = null;
        continue;
    }
    if ($carg === '--delimiter') {
        if (strlen($delimiter) > 1) {
            dieNice(5, 'Delimiter must be a single character only.');
        }
        $delimiter = $arg;
        $carg = null;
        continue;
    }
    dieNice(1, 'invalid argument ' . $arg);
}
// End of handling the command line arguments.

// Start working here...
if (!$infile) {
    dieNice(1, 'No Moodle XML File provided.');
}
if (!file_exists($infile)) {
    dieNice(2, 'File ' . $infile . ' does not exist.');
}
if (!$data = file_get_contents($infile)) {
    dieNice(2, 'Could not read file ' . $infile);
}
if (!$xml = simplexml_load_string($data)) {
    dieNice(3, 'Could not parse XML from file ' . $infile);
}
if (!$outfile) {
    $p = strrpos($infile, '.');
    if ($p === false) {
        $p = strlen($infile);
    }
    $outfile = substr($infile, 0, $p) . '.csv';
}
$fout = fopen($outfile, 'wb');
if (!$fout) {
    dieNice(4, 'Could not open ' . $outfile . ' for writing');
}

$cols = [];
$csvdata = [];

foreach ($xml->xpath('//question') as $question) {
    handler($question);
}

fputcsv($fout, $cols, $delimiter);
foreach ($csvdata as $line) {
    while (count($line) < count($cols)) {
        $line[] = '';
    }
    fputcsv($fout, $line, $delimiter);
}
fclose($fout);
// Finshed working here.

/**
 * Get column number for a column name.
 * @param string $name
 * @return int
 */
function getColIdx(string $name): int {
    global $cols;
    foreach ($cols as $i => $col) {
        if ($col === $name) {
            return $i;
        }
    }
    $cols[] = $name;
    return count($cols) - 1;
}

/**
 * Format/strip html tags from a value.
 * @param string $value
 * @return string
 */
function formatHtml(string $value): string {
    global $striptags, $keeptags;
    if ($striptags) {
        return strip_tags($value, $keeptags);
    }
    return $value;
}

/**
 * Handler function that (in this case) handles the mc questions.
 * @param $question
 * @return void
 */
function handler($question) {
    global $csvdata;
    if ((string)$question['type'] === 'multichoice') {
       $row = [];
       $row[getColIdx('name')] = trim($question->name->text ?? '');
       $row[getColIdx('questiontext')] = formatHtml((string)$question->questiontext->text ?? '');
       $row[getColIdx('generalfeedback')] = formatHtml((string)$question->generalfeedback->text ?? '');
       $row[getColIdx('defaultgrade')] = (float)$question->defaultgrade ?? 0.0;
       $row[getColIdx('penalty')] = (float)$question->penalty ?? 0.0;
       $row[getColIdx('hidden')] = (int)$question->hidden ?? 0;
       $row[getColIdx('idnumber')] = (string)$question->idnumber ?? '';
       $row[getColIdx('single')] = ((string)$question->single ?? 'true') === 'true' ? 1 : 0;
       $row[getColIdx('shuffleanswers')] = ((string)$question->shuffleanswers ?? 'true') === 'true' ? 1 : 0;
       $row[getColIdx('answernumbering')] = (string)$question->answernumbering ?? '';
       $row[getColIdx('showstandardinstruction')] = (int)$question->showstandardinstruction ?? 0;
       $row[getColIdx('correctfeedback')] = formatHtml((string)$question->correctfeedback->text ?? '');
       $row[getColIdx('partiallycorrectfeedback')] = formatHtml((string)$question->partiallycorrectfeedback->text ?? '');
       $row[getColIdx('incorrectfeedback')] = formatHtml((string)$question->incorrectfeedback->text ?? '');
       $i = 1;
       foreach($question->answer as $answer) {
           $row[getColIdx('answer_' . $i)] = formatHtml((string)$answer->text ?? '');
           $row[getColIdx('answer_' . $i . '_fraction')] = (float)$answer['fraction'] ?? 0.0;
           $row[getColIdx('answer_' . $i . '_feedback')] = formatHtml((string)$answer->feedback->text ?? '');
           $i++;
       }
       $csvdata[] = $row;
    }
}

/**
 * Terminate the script with an exit code.
 * @param int $code
 * @param string|null $msg
 * @return void
 */
function dieNice(int $code = 0, ?string $msg = null) {
    if ($code > 0) {
        echo $msg . PHP_EOL;
        echo 'See --help for more details' . PHP_EOL;
    }
    exit($code);
}

File: mc2csv.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

"""
Takes a Moodle XML export file of questions and exports all the
multiple choice questions into a csv file.

Usage php mc2csv.php --in questions.xml

Optional arguments:
--delimiter char    Delimiter character for the csv file, default is ";"
--keephtml          Do not strip html tags at all.
--keeptags 'a,b,c'  List of tags that should not be eliminated when
                    stripping html from the questions. The default
                    tags that are not purged: i,img,s,strong,sub,sub,u
--out filename.csv  if not provided the base name of the xml file is
                    used, trailed by a .csv suffix.
"""

import xml.etree.ElementTree as ET
from html.parser import HTMLParser
import os, sys, csv

def dieNice(errMsg = ""):
    print("Error: {0}\nUsage: {1} --in moodle_questions.xml\nType --help for more details."
        .format(errMsg, os.path.basename(sys.argv[0])))
    sys.exit(1)

class HTMLTagRemover(HTMLParser):
    def __init__(self, keeptags):
        super().__init__()
        self.result = []
        self.keeptags = keeptags

    def handle_data(self, data):
        self.result.append(data)

    def handle_starttag(self, tag, attrs):
        try:
            self.keeptags.index(tag)
            self.result.append('<' + tag + ', '.join(['{}="{}"'.format(k,v) for k,v in attrs.iteritems()]) + '>')
        except:
            pass

    def handle_endtag(self, tag):
        try:
            self.keeptags.index(tag)
            self.result.append('</' + tag + '>')
        except:
            pass

    def handle_startendtag(tag, attrs):
        try:
            self.keeptags.index(tag) > -1
            self.result.append('<' + tag + ', '.join(['{}="{}"'.format(k,v) for k,v in attrs.iteritems()]) + '/>')
        except:
            pass

    def get_text(self):
        return ''.join(self.result)


class McHandler:

    def __init__(self, cols, options):
        self.cols = cols
        self.keeptags = options['keeptags']
        self.keephtml = options['keephtml']

    def insertColInRow(self, row, name, value):
        try:
            i = self.cols.index(name)
            row[i] = value
        except ValueError:
            self.cols.append(name)
            row.append(value)

    def formatHtml(self, value):
        if value == None:
            return ''
        if self.keephtml == True:
            return value
        remover = HTMLTagRemover(self.keeptags)
        remover.feed(value)
        return remover.get_text()

    def handle(self, question):
        if 'type' not in question.attrib or question.attrib['type'] != 'multichoice':
            return []
        row = ['' for i in range(0, len(self.cols))]
        self.insertColInRow(row, 'name', question.find('name/text').text)
        self.insertColInRow(row, 'questiontext', self.formatHtml(question.find('questiontext/text').text))
        self.insertColInRow(row, 'generalfeedback', self.formatHtml(question.find('generalfeedback/text').text))
        self.insertColInRow(row, 'defaultgrade', question.find('defaultgrade').text)
        self.insertColInRow(row, 'penalty', question.find('penalty').text)
        self.insertColInRow(row, 'hidden', question.find('hidden').text)
        self.insertColInRow(row, 'idnumber', question.find('idnumber').text)
        self.insertColInRow(row, 'single', '1' if question.find('single').text.strip() == 'true' else '0')
        self.insertColInRow(row, 'shuffleanswers', '1' if question.find('shuffleanswers').text.strip() == 'true' else '0')
        self.insertColInRow(row, 'answernumbering', question.find('answernumbering').text)
        self.insertColInRow(row, 'showstandardinstruction', question.find('showstandardinstruction').text)
        self.insertColInRow(row, 'correctfeedback', self.formatHtml(question.find('correctfeedback/text').text))
        self.insertColInRow(row, 'partiallycorrectfeedback', self.formatHtml(question.find('partiallycorrectfeedback/text').text))
        self.insertColInRow(row, 'incorrectfeedback', self.formatHtml(question.find('incorrectfeedback/text').text))
        i = 1
        for answer in question.findall('answer'):
            self.insertColInRow(row, 'answer_' + str(i), self.formatHtml(answer.find('text').text))
            self.insertColInRow(row, 'answer_' + str(i) + '_fraction', answer.attrib['fraction'])
            self.insertColInRow(row, 'answer_' + str(i) + '_feedback', self.formatHtml(answer.find('feedback/text').text))
            i = i + 1

        return row

class QuestionXml:

    def __init__(self):
        self.infile = None
        self.outfile = None
        self.delimiter = ';'
        self.keeptags = ['i', 'img', 's', 'strong', 'sub', 'sub', 'u']
        self.keephtml = False
        self.cols = []
        self.rows = []


    def process(self):
        # Passing the path of the
        # xml document to enable the
        # parsing process
        if self.infile == None:
            dieNice('No Moodle XML file provided')

        try:
            tree = ET.parse(self.infile)
        except FileNotFoundError:
            dieNice('File {0} not found'.format(self.infile))
        except ET.ParseError:
            dieNice('Error parsing XML in file ' + self.infile)

        # getting the parent tag of
        # the xml document
        root = tree.getroot()

        handler = McHandler(self.cols, {'keephtml': self.keephtml, 'keeptags': self.keeptags})
        for question in root.findall('question'):
            row = handler.handle(question)
            if len(row) > 0:
                self.rows.append(row)


    def writeCsv(self):
        # Check, if we have a file name set.
        if self.outfile == None:
            p = self.infile.rfind('.')
            if p == -1:
                p = len(self.infile)
            self.outfile = self.infile[0, p] + '.csv'
        # Open the file.
        try:
            fp = open(self.outfile, "w")
        except:
            dieNice('Could not open file %s for writing result' % self.outfile)
        # Setup the writer and put the columns (as header) and then all rows into the file.
        writer = csv.writer(fp, delimiter = self.delimiter)
        writer.writerow(self.cols)
        for row in self.rows:
            writer.writerow(row)
        fp.close()


def main():
    """Evaluate the cli arguments, built up the work object
    with the parameters to process the xml file. Finally write
    the output into a csv file."""

    # available options that can be changed via the command line
    options = ['in', 'out', 'delimiter', 'keephtml', 'keeptags', 'help']

    # the worker that does the parsing.
    worker = QuestionXml()

    # try to fetch the command line args
    currentCmd = ''
    for i in range(len(sys.argv)):
        if i == 0:
            continue
        arg = sys.argv[i]
        if arg[0:2] == '--':
            currentCmd = arg[2:]
            if not(currentCmd in options):
                dieNice(f"Invalid argument {0}".format(currentCmd))
            if currentCmd == 'help':
                print(__doc__)
                sys.exit(0)
            elif currentCmd == 'keephtml':
                worker.keephtml = True
                currentCmd = ''
        elif len(currentCmd) > 0:
            if currentCmd == 'in':
                worker.infile = arg
            elif currentCmd == 'out':
                worker.outfile = arg
            elif currentCmd == 'delimiter':
                if len(arg) > 1:
                    dieNice('Delimiter must be one character only')
                worker.delimiter = arg
            elif currentCmd == 'keeptags':
                worker.keeptags = list(map(lambda x: x.strip(), arg.split(',')))
            currentCmd = ''
        else:
            dieNice('invalid argument: ' + arg)

    # process the data now
    worker.process()
    worker.writeCsv()

if __name__ == "__main__":
    main()
